### ISSUE TITLE
Real world cases where Go codes that generated by re2go is magnitude slower than Go's standard regexp

### ISSUE THREAD

**Issue by RadhiFadlillah** [skvadrik/re2c#487]
## Background

Go's regexp is known to be [slow](https://github.com/golang/go/issues/26623#issuecomment-1899965448). So, to improve performance sometimes I use re2go to compile the regex into Go code.

However, I found several cases where code generated by re2go is magnitude slower than Go's standard library. In the worst case, code by re2go take around 9s (yes, it's in seconds) while Go's standard library only take 15ms.

Here I use the latest stable version of re2go, which is v3.1.

## Code Sample

I've made a [repository](https://github.com/RadhiFadlillah/re2go-slow-issue) that contains code and web pages where this issue occurs. Here is the output of code in that repository when run on my Linux PC with Intel i7-8550U and RAM 16 GB:

```
File "sample/01.html"
  re2go: found 0 matches in 9249 ms
  std  : found 0 matches in 15 ms
  re2go is 616x slower than std

File "sample/02.html"
  re2go: found 0 matches in 7570 ms
  std  : found 0 matches in 14 ms
  re2go is 540x slower than std

File "sample/03.html"
  re2go: found 0 matches in 2030 ms
  std  : found 0 matches in 13 ms
  re2go is 156x slower than std

File "sample/04.html"
  re2go: found 0 matches in 1051 ms
  std  : found 0 matches in 16 ms
  re2go is 65x slower than std
```

## Summary of the Sample

I believe the code in that repository is readable and easy enough to understand, however for clarity here I also put a short summary about that repository.

In short, I'm using the following regex pattern to find emails in web pages:

```
[\+\-\.0-9A-Z_a-z]+@[\-\.0-9A-Z_a-z]+[\.][\-\.0-9A-Z_a-z]+
```

For Go's standard library, the regex is generated simply by running `regexp.MustCompile`:

```go
var rxEmail = regexp.MustCompile(`[\+\-\.0-9A-Z_a-z]+@[\-\.0-9A-Z_a-z]+[\.][\-\.0-9A-Z_a-z]+`)
```

For re2go, the code is generated by following `.re` template:

```go
package main

func re2Email(bytes []byte) int {
	var count int
	var cur, mar int
	lim := len(bytes)

	// Peek function
	peek := func(bytes []byte, cur int, lim int) byte {
		if cur < lim {
			return bytes[cur]
		}
		return 0
	}

	for { /*!re2c
		re2c:eof = 0;
		re2c:yyfill:enable = 0;
		re2c:posix-captures = 0;
		re2c:case-insensitive = 0;
		re2c:define:YYCTYPE     = byte;
		re2c:define:YYPEEK      = "peek(bytes, cur, lim)";
		re2c:define:YYSKIP      = "cur += 1";
		re2c:define:YYBACKUP    = "mar = cur";
		re2c:define:YYRESTORE   = "cur = mar";
		re2c:define:YYLESSTHAN  = "lim <= cur";
		re2c:define:YYSTAGP     = "@@{tag} = cur";
		re2c:define:YYSTAGN     = "@@{tag} = -1";
		re2c:define:YYSHIFTSTAG = "@@{tag} += @@{shift}";

		email = [\+\-\.0-9A-Z_a-z]+@[\-\.0-9A-Z_a-z]+[\.][\-\.0-9A-Z_a-z]+;

		{email} { count += 1; continue }
		*       { continue }
		$       { return count }
		*/
	}
}
```

Which then compiled by re2go using following command:

```
re2go -F --input-encoding utf8 --utf8 --no-generation-date -i $INPUT -o $OUTPUT
```

The generated code then compared with Go's standard regexp library using several web pages in [sample directory](https://github.com/RadhiFadlillah/re2go-slow-issue/tree/master/sample) of that repository.

## Closing

So, what do you think? Did I do something wrong? Are there anything that I can do in my side to improve the performance?

Thanks!

**Comment by RadhiFadlillah** [comment]
I've updated the repository to use sentinel with bound checks for handling the end of input.

To my surprise, sentinel with bound checks is faster than handling the end of input with custom check. It's still slower than Go's standard library though.

```
File "sample/01.html"
  Go std regex        : found 0 matches in 20 ms
  re2go custom check  : found 0 matches in 9697 ms (484x slower)
  re2go sentinel bound: found 0 matches in 7677 ms (383x slower)

File "sample/02.html"
  Go std regex        : found 0 matches in 15 ms
  re2go custom check  : found 0 matches in 7638 ms (509x slower)
  re2go sentinel bound: found 0 matches in 6200 ms (413x slower)

File "sample/03.html"
  Go std regex        : found 0 matches in 13 ms
  re2go custom check  : found 0 matches in 1918 ms (147x slower)
  re2go sentinel bound: found 0 matches in 1471 ms (113x slower)

File "sample/04.html"
  Go std regex        : found 0 matches in 17 ms
  re2go custom check  : found 0 matches in 1063 ms (62x slower)
  re2go sentinel bound: found 0 matches in 736 ms (43x slower)
```

**Comment by skvadrik** [comment]
Hi, thanks for reporting this interesting case!

> To my surprise, sentinel with bound checks is faster than handling the end of input with custom check.

This is expected: with custom checks the lexer performs a check on every character, which is very inefficient (it is documented [here](http://re2c.org/manual/manual_go.html#custom-checks)).

I can confirm that re2go code is much slower. I did a few modifications to clear unused re2go definitions, fix re2go warnings and use sentinel with bound checks instead of custom checks, but the result is roughly the same.

I think I know the reason. I made the following experiment.

If I replace `z` with `y` in both regexps, effectively reducing the range by one character (`[+\-.0-9A-Z_a-y]+[@][\-.0-9A-Z_a-z]+[.][\-.0-9A-Z_a-z]+`), re2go generated code becomes fast. In fact, it can be not necessarily `z` but a different letter or a digit that is taken out of the first range, and the result is similar. The difference in re2go generated code is minor (as one would expect), and the difference in the compiled assembly code is minor (I looked at it using `objdump -DR main` and then `diff` the `re2Email` functions). `perf stat` shows a lot more instructions executed and branches taken in the slow case, so it seems that the slow case just does more work.

Looking further, there're a few very long base64-encoded strings in the input text. If I cut them out, re2go code becomes fast.

So my conclusion is, re2c spends a long time in a vain attempt to recognize those strings as emails, and it finally fails after wasting much effort. And then it skips one character (just one!) and starts all over again from the next character. What a waste!!! It's quadratic. That goes along with the experiment: removing one character makes re2c fail much faster, as this character statistically occurs frequently enough in any base64 encoded string.

Go regexp probably uses a different algorithm in this case; I think it uses something like [Rabin-Karp algorithm](https://en.wikipedia.org/wiki/Rabin%E2%80%93Karp_algorithm) that is designed to quickly filter out positions of the text that cannot match the pattern, and it shines in a case like this.

In practice what I'd do with re2go to make it faster is to add a rule that skips such strings:
```
    "base64," [+\-.0-9A-Z_a-z]+ { continue } // skip very long base64-encoded strings
```
That makes re2go code very fast, and it's something that often needs to be done in practical lexers, as one has to skip comments and strings that don't contain useful lexemes.

Of course, it would be better if re2go was smarter and applied something like Rabin-Karp algorithm, but I'm not sure it's possible with ahead of time compilation (it may require the input string - I need to refresh my memory on these fast string search algorithms).

Hope that helps!

**Comment by skvadrik** [comment]
Just a note that if you apply the extra rule I suggested, please double-check what characters are allowed in base64 as I just picked the range from the email regexp.

**Comment by RadhiFadlillah** [comment]
Thanks for the response!

> thanks for reporting this interesting case!

You are welcome! You might be interested with my benchmarks that compare various Go's regex packages (including the ones that use external regex like re2 and Hyperscan):

- [This one](https://github.com/RadhiFadlillah/go-regex-benchmark) uses input from concatenation of texts in [Learn X in Y minutes](https://github.com/adambard/learnxinyminutes-docs) repository.
- [This one](https://github.com/RadhiFadlillah/go-regex-benchmark-web) is similar with the above, but uses input from around 1,000 web pages so it might be more suitable for real world benchmark.

> ... fix re2go warnings ...

I just realized re2go by default doesn't print any warnings. Thanks for the hint!

> In practice what I'd do with re2go to make it faster is to add a rule that skips such strings

Unfortunately it's not suitable for my case. I have a Go project that parses several thousand web pages and heavily use regex. I'm worried if an extra rule for a web page might be not suitable for the other pages.

But thanks for the suggestions, it might be suitable for my other projects.

> So my conclusion is, re2go code spends a long time in a vain attempt to recognize those strings as emails, and it finally fails after wasting much effort. And then it skips one character (just one!) and starts all over again from the next character.

Yes, this is it! I've put some logging in my codes and could confirm there are several wasteful backtracking:

```
CURSOR 0 => '<'
CURSOR 1 => '!'
CURSOR 2 => 'D'		match candidate in cursor 2
CURSOR 3 => 'O'
CURSOR 4 => 'C'
CURSOR 5 => 'T'
CURSOR 6 => 'Y'
CURSOR 7 => 'P'
CURSOR 8 => 'E'
CURSOR 9 => ' '		pattern doesn't match, backtrack
CURSOR 3 => 'O'		continue to cursor 3
CURSOR 4 => 'C'
CURSOR 5 => 'T'
CURSOR 6 => 'Y'
CURSOR 7 => 'P'
CURSOR 8 => 'E'
CURSOR 9 => ' '		pattern doesn't match, backtrack
CURSOR 4 => 'C'		continue to cursor 4
CURSOR 5 => 'T'
CURSOR 6 => 'Y'
CURSOR 7 => 'P'
CURSOR 8 => 'E'
CURSOR 9 => ' '		pattern doesn't match, backtrack
CURSOR 5 => 'T'		continue to cursor 5
...
```

> Go regexp probably uses a different algorithm in this case;

From what I understand, Go regexp uses NFA algorithm without backtracking and uses one-pass NFA whenever possible. It's developed by Russ Cox that develops [`google/re2`](https://github.com/google/re2) which implements converting NFA to DFA. There was a [plan](https://github.com/golang/go/issues/11646) to implement that DFA converters to Go, but it seems to be halted.

The method Go uses to avoid backtracking is explained in Russ [`regexp1`](https://swtch.com/~rsc/regexp/regexp1.html) blog. I'm sure you've read it, but here is its summary for other people that might visit this issue:

> One way to simulate perfect guessing is to guess one option, and if that doesn't work, try the other. For example, consider the NFA for `abab|abbb` run on the string `abbb`: 
>
> ![NFA diagram](https://github.com/user-attachments/assets/c903fd53-edee-4175-9202-8b00d5bbb5b3)
>
> ![NFA with backtracking](https://github.com/user-attachments/assets/db34a470-eb9c-4732-a35c-88ed22bed543)
>
> ...
>
> A more efficient but more complicated way to simulate perfect guessing is to guess both options simultaneously. In this approach, the simulation allows the machine to be in multiple states at once. To process each letter, it advances all the states along all the arrows that match the letter.
>
> ![NFA without backtracking](https://github.com/user-attachments/assets/7c518181-4837-4abb-914a-d6078771600d)
>
> ... This is a dramatic improvement over the exponential time required by the backtracking approach ...

Since it's still a variant of NFA algorithm, I wonder if it could be implemented by re2go.

**Comment by RadhiFadlillah** [comment]
> Codesearch gives lower counts than others, could it be that it does not find all occurrences?

Yes, codesearch only works like grep, so it returns the line containing matches and not the matching phrases itself. That's why its count is lower than the others. I included it just to see how fast Go regexp might be in the future.

> Besides, the added rule cannot be confused with other rules, so it cannot break anything.

Thanks, now it's clicked on me! I'm still wrapping my head around re2go sometimes I forget that it supports multiple regexes in one .re template. So from what I understand, now we use two regexes:

1. One is `base64 regex` to filter base64 patterns from the web page.
2. The other one is the `email regex` to catch the actual email patterns. Since the long base64 in the web page will be excluded by the 1st regex, this 2nd regex will perform way faster.

> re2go is DFA-based; making it use NFA will only decrease performance. Using sublinear algorithms would be good, but I think they require the input string, which is not available for ahead-of-time generators like re2go.

Thanks for the explanations! Besides adding additional regex to filter the page, are there any other tips for me to improve the performance? Are there any flags or options that I need to change? Or are there any tips on debugging the bottleneck if I meet case like this again?

Thanks!

**Comment by skvadrik** [comment]
> I'm still wrapping my head around re2go sometimes I forget that it supports multiple rules in one .re template. So from what I understand, now we use two rules:
>
>    1. The first rule is base64 regex to filter base64 patterns from the web page.
>    2. The second rule is the email regex to catch the actual email patterns. Since the long base64 in the web page will be excluded by the 1st rule, this 2nd rule will perform way faster.

Correct. Note that all the rules are matched at once: re2go merges them together into one DFA with a separate exit state for each rule. The longest match wins, and if there are many equal-length matches, the one listed earlier in the code wins. In our case there are two matching rules: the one for base64 (it matches a long string) and the default rule `*` (it matches just one character). The long rule wins, and the lexer consumes the whole string. If there were no base64 rule, the lexer would still attempt to match the long string as part of an email address, but then it would find no `@` and fallback to the short default rule `*`.

By the way, the order of the rules does not matter here: you can put them in any order. Default rule will always have the lowest priority, and base64/email rules do not overlap.

> Besides adding additional rules to filter the page, are there any other tips for me to improve the performance? Are there any flags or options that I need to change? Or are there any tips on debugging the bottleneck if I meet case like this again?

Try `--bit-vectors` option, it helps a bit in this case. It requires using a different end-of-input check (bounds checking), here's how the modified re2email.re looks like:
```go
package main

import "bytes"

/*!max:re2c*/

func findEmails(input []byte) int {
        var count int
        var cur, mar int
        input = append(input, bytes.Repeat([]byte{0}, YYMAXFILL)...)
        lim := len(input)

        for { /*!re2c
                re2c:bit-vectors = 1;
                re2c:define:YYCTYPE    = byte;
                re2c:define:YYPEEK     = "input[cur]";
                re2c:define:YYSKIP     = "cur += 1";
                re2c:define:YYBACKUP   = "mar = cur";
                re2c:define:YYRESTORE  = "cur = mar";
                re2c:define:YYLESSTHAN = "lim - cur < @@";
                re2c:define:YYFILL     = "panic(`fill unexpected`)";

                email = [\+\-\.0-9A-Z_a-z]+@[\-\.0-9A-Z_a-z]+[\.][\-\.0-9A-Z_a-z]+;
                base64 = [;]base64[,][+0-9A-Za-z/]+;

                {email}  { count += 1; continue }
                {base64} { continue } // skip very long base64-encoded strings
                [\x00]   { return count } // the end
                *        { continue }
                */
        }
}
```

Generally, the first thing to worry is to reduce end of input checks as they kill performance, and this means trying sentinel/bounds checks methods or their combination. As for debugging bottlenecks, there's no easy strategy but you should inspect rules that have iteration in them, as they are always greedy and will try to match as long as possible.

**Comment by skvadrik** [comment]
Also, a few notes regarding your re2go lexers (not affecting performance):

- It's better to use re2go native syntax instead of the `-F` option, as the latter is more like a legacy mode that remains from the early days of re2c, and it has some annoying limitations and diversions from real Flex syntax.

- You don't need `--input-encoding utf8` (the only reason it is used is `-F` option that will require escaping `@` with brackets if not used in UTF8 mode).

**Comment by RadhiFadlillah** [comment]
I've updated the [issue repository](https://github.com/RadhiFadlillah/re2go-slow-issue) to include more `.re` templates and more sample files. In one of the template I've added another rule to also exclude chars that too long, which is impossible to be email:

```
tooLong = [+\-.0-9A-Z_a-z]{100,};
```

> Try `--bit-vectors` option, it helps a bit in this case.

Unfortunately, when I tried it in my [real world benchmark](https://github.com/RadhiFadlillah/re2go-slow-issue) that option actually makes the performance slower. In fact from 1,058 web pages in that benchmark only one web page that become faster while the others become slower. The difference are not too far though, at worst it's only around 20 milliseconds.

In case you're interested, I've attached CSV file that list the name of files and its speed [here](https://github.com/user-attachments/files/16979935/standard-vs-bitvector.csv). For your convenience, I've added some of those files as sample in the issue repository. Here is the related output:

```
File "sample/05.html"
  Go std regex     : 8,585 µs
  re2go custom     : 4,371 µs (1x faster)
  re2go sentinel   : 3,261 µs (2x faster)
  re2go sentinel BV: 493 µs (17x faster)
  re2go 2 rules    : 3,079 µs (2x faster)
  re2go 2 rules BV : 509 µs (16x faster)
  re2go 3 rules    : 3,301 µs (2x faster)
  re2go 3 rules BV : 689 µs (12x faster)

File "sample/06.html"
  Go std regex     : 8,690 µs
  re2go custom     : 12,951 µs (1x slower)
  re2go sentinel   : 9,488 µs (1x slower)
  re2go sentinel BV: 6,960 µs (1x faster)
  re2go 2 rules    : 9,086 µs (1x slower)
  re2go 2 rules BV : 7,932 µs (1x faster)
  re2go 3 rules    : 2,674 µs (3x faster)
  re2go 3 rules BV : 3,838 µs (2x faster)

File "sample/07.html"
  Go std regex     : 78,352 µs
  re2go custom     : 55,233 µs (1x faster)
  re2go sentinel   : 31,518 µs (2x faster)
  re2go sentinel BV: 27,455 µs (2x faster)
  re2go 2 rules    : 24,278 µs (3x faster)
  re2go 2 rules BV : 25,478 µs (3x faster)
  re2go 3 rules    : 23,033 µs (3x faster)
  re2go 3 rules BV : 36,690 µs (2x faster)
```

> It's better to use re2go native syntax instead of the -F option.

Sure, I'll keep it in my mind. Unfortunately right now I'm too used with Go's regex syntax which is kinda similar as the Flex syntax.

**Comment by skvadrik** [comment]
> In one of the template I've added another rule to also exclude chars that too long, which is impossible to be email

A word of caution: it is generally not a good idea to use high bounded repetition counters (100 in this case), as the generated DFA quickly becomes too large. In this case it may have a positive effect on performance as it unrolls the first 100 iterations of the loop, which makes some other optimizations possible, but generally it's discouraged and not practical.

> Unfortunately, when I tried it in my [real world benchmark](https://github.com/RadhiFadlillah/re2go-slow-issue) that option actually makes the performance slower.

Is the link incorrect by any chance? Shouldn't it be https://github.com/RadhiFadlillah/go-regex-benchmark?

It is expected that `--bit-vectors` may be sometimes slower and sometimes faster depending on each use case, so it's worth trying and setting it on per-file basis. I recommended it because when I experimented with your repo https://github.com/RadhiFadlillah/go-regex-benchmark I saw a noticeable improvement from roughly 5.8s to 4.6s (with the re2email.re listed above, 5.8s when I commented out the line `re2c:bit-vectors = 1;`).

If you intend to integrate your fixes (mainly the addition of the extra rule) into the real-world benchmark, let me know and I will check afterwards if I can spot any further improvements. And thanks a lot for all the detailed info you provided, it is much appreciated.

---

Also, just FYI, re2go allows to avoid code duplication.

Suppose you have two almost identical blocks that differ in just a few configurations:
```
/*!re2c
    // ... code
*/

/*!re2c
    // ... code
    extra-config = 1;
*/
```
You can rewrite it as follows (where `X` is an arbitrary name):
```
/*!rules:re2c:X
   // ... code
*/

/*!use:re2c:X*/

/*!use:re2c:X
    extra-config = 1;
*/
```
More details [here](http://re2c.org/manual/manual_go.html#reusable-blocks). re2go also allows you to move common definitions to [include files](http://re2c.org/manual/manual_go.html#include-files).

**Comment by skvadrik** [comment]
I looked at the samples and I have [what I think is] a good idea.

We have a problem with base64-encoded strings of different form. So, instead of having a few specialized extra rules for different html tags, we should have one generic simple rule that skips all strings that begin like email, but don't end like email:
```
    [+0-9A-Za-z]+; { continue }
```
If this rule matches the beginning of an email (and it will), then the longer overlapping rule for email will definitely take precedence, and no harm done. If this is, however, not an email, then this rule will quickly skip over it. (Note that I removed forward slash from the base64 range, so that this greedy iteration is a strict subset of email.)

I tried this on the real world benchmark, and I confirm that the counter results are the same, but performance is better.

**Comment by RadhiFadlillah** [comment]
> Is the link incorrect by any chance?

Yes I'm sorry, that link is wrong. The correct one is https://github.com/RadhiFadlillah/go-regex-benchmark-web.

> It is expected that `--bit-vectors` may be sometimes slower and sometimes faster depending on each use case, so it's worth trying and setting it on per-file basis.

In that case, I guess I'll disable it for now since the code generated using `--bit-vector`  is a bit hard to follow compared to the normal one.

> Also, just FYI, re2go allows to avoid code duplication. If you have two almost identical blocks that differ in just a few configurations or rules, you can extract the common part into a reusable block.

I haven't finished reading the manual, so thanks for the hint! Right now I only reached the submatch extraction, one of these days I'll finish reading it.

> ...we should have one generic simple rule that skips all strings that begin like email, but don't end like email:
>
> ...
>
> If this rule matches the beginning of an email (and it will), then the longer overlapping rule for email will definitely take precedence, and no harm done. If this is, however, not an email, then this rule will quickly skip over it.

In that case, since the email pattern looks like this:

```
email = [+\-.0-9A-Z_a-z]+@[\-.0-9A-Z_a-z]+[.][\-.0-9A-Z_a-z]+
```

Then shouldn't the filter use the entire prefix of the email pattern before '@':

```
prefix = [+\-.0-9A-Z_a-z]+
```

**Comment by RadhiFadlillah** [comment]
> Then shouldn't the filter use the entire prefix of the email pattern before '@':

I think it might works. Our initial problem was:

> re2go code spends a long time in a vain attempt to recognize those strings as emails, and it finally fails after wasting much effort. And then it skips one character (just one!) and starts all over again from the next character.

For example, let's say I have a short string like this:

```
Contact me in john@doe.com
```

In my initial post I use the following `.re` templates (omitted for clarity):

```go
func re2Email(bytes []byte) int {
	for { /*!re2c
		email = [\+\-\.0-9A-Z_a-z]+@[\-\.0-9A-Z_a-z]+[\.][\-\.0-9A-Z_a-z]+;

		{email} { count += 1; continue }
		*       { continue }
		$       { return count }
		*/
	}
}
```

That template will works like this:

```
0	Email me in john@doe.com	Store marker in 0.
	^
1	Email me in john@doe.com
	 ^
2	Email me in john@doe.com
	  ^
3	Email me in john@doe.com
	   ^
4	Email me in john@doe.com
	    ^
5	Email me in john@doe.com	We hit space, so it's not email. Go back.
	     ^
1	Email me in john@doe.com	Store marker in 1.
         ^
2	Email me in john@doe.com
	  ^
3	Email me in john@doe.com
	   ^
4	Email me in john@doe.com
	    ^
5	Email me in john@doe.com	We hit space, so it's not email. Go back.
	     ^
2	Email me in john@doe.com	Store marker in 2.
	  ^
```

As you can see, not only do we have to go far back, we can only go one step forward. Now I added a `prefix` rule like this:

```go
func re2Email(bytes []byte) int {
	for { /*!re2c
		email  = [\+\-\.0-9A-Z_a-z]+@[\-\.0-9A-Z_a-z]+[\.][\-\.0-9A-Z_a-z]+;
		prefix = [\+\-\.0-9A-Z_a-z]+;

		{email}  { count += 1; continue }
		{prefix} { continue }
		*        { continue }
		$        { return count }
		*/
	}
}
```

With that changed, now the template will works like this:

```
0	Email me in john@doe.com	Store marker in 0.
	^
1	Email me in john@doe.com
	 ^
2	Email me in john@doe.com
	  ^
3	Email me in john@doe.com
	   ^
4	Email me in john@doe.com
	    ^
5	Email me in john@doe.com	We hit space so it's not email but only prefix. Continue.
	     ^
6	Email me in john@doe.com
	      ^
7	Email me in john@doe.com
	       ^
8	Email me in john@doe.com
	        ^
9	Email me in john@doe.com
	         ^
...
```

As you can see, now the generated code is only going forward so it's way faster. I've updated both of my benchmarks, and re2go now is magnitude faster, almost as fast as re2. The counter result are the same as well.

So what do you think? Is it safe, or are there any catch that I haven't found?

Thanks!

**Comment by skvadrik** [comment]
> In that case, since the email pattern looks like this [...] Then shouldn't the filter use the entire prefix of the email pattern before '@'

Yes, it should. I used a smaller range `[+0-9A-Za-z]` because I took the intersection of base64 range and email range (so it was enough to fix the problem with base64 strings), but it's better to use the full email range `[+\-.0-9A-Z_a-z]`. You may give them names like this:

```
    suffix_char = [\-.0-9A-Z_a-z];
    prefix_char = {suffix_char}|[+];
    suffix = {suffix_char}+;
    prefix = {prefix_char}+;
    email = {prefix}@{suffix}[.]{suffix};
```

> As you can see, now the generated code is only going forward so it's way faster. I've updated both of my benchmarks, and re2go now is magnitude faster, almost as fast as re2. The counter result are the same as well.

Good \o/

> So what do you think? Is it safe, or are there any catch that I haven't found?

There's no catch, it is safe. As long as the range in the prefix rule is a subset of the email range (exactly the same, or with some characters taken out), prefix rule cannot consume any valid email (because the longer email rule will take precedence). It would be unsafe if you included some other characters in the prefix rule that are not present in email: in that case prefix rule might have consumed parts of valid emails together with non-emails.

But the current rule is definitely safe -- the only thing I'd do is to add a comment explaining that it's needed for performance.

**Comment by RadhiFadlillah** [comment]
> There's no catch, it is safe.

Nice, that's good to hear!

> As long as the range in the prefix rule is a subset of the email range (exactly the same, or with some characters taken out)

In that case, since the email regex has three quantifier groups:

![regex visualization](https://github.com/user-attachments/assets/4e93cab8-6e46-49b7-955f-941faa39ae5d)

To be more safe I think I can go further and split it into three rules:

```go
func re2Email(bytes []byte) int {
	for { /*!re2c
		quant1 = [+\-.0-9A-Z_a-z]+;
		quant2 = [+\-.0-9A-Z_a-z]+@[\-.0-9A-Z_a-z]+;
		email  = [+\-.0-9A-Z_a-z]+@[\-.0-9A-Z_a-z]+[.][\-.0-9A-Z_a-z]+;

		{email}  { count += 1; continue }
		{quant1} { continue }
		{quant2} { continue }
		*        { continue }
		$        { return count }
		*/
	}
}
```

At this point I understand why this problem happened and how to fix it, so I'll close this issue. Thank you for your help and for your patience in answering my questions.

Have a nice day!

**Comment by skvadrik** [comment]
> To be more safe I think I can go further and split it into three rules.

This is in fact not as good as what we had before. The second range `[\-.0-9A-Z_a-z]` contains dot `.`, so any valid email would be matched by `{quant2}`. It's only the rule order now that makes `{email}` match before `{quant2}`. And `{quant2}` is not very useful from performance point of view as well, because the strings it matches are very unlikely in HTML (something like `john@doe` not followed by `.com`), so it's not worth making the lexer more complex for such a rare case. My recommendation is to keep the old prefix rule.

> At this point I understand why this problem happened and how to fix it, so I'll close this issue. Thank you for your help and for your patience in answering my questions.

And thank you for this interesting case!

**Comment by falco467** [comment]
I think you could vastly speed up this re2c search, by discarding all character-runs which don't contain an @ and end with any non-mail allowed character. Discard: `[^@]+[^\+\-\.0-9A-Z_a-z]`
